src/vmprof_main.h

```diff
@@ -199,6 +239,21 @@ static void sigprof_handler(int sig_nr, siginfo_t* info, void *ucontext)
     // get_current_thread_state returns a sane result
     while (__sync_lock_test_and_set(&spinlock, 1)) {
     }
+
+#ifdef VMPROF_UNIX
+    // SIGNAL ABUSE AHEAD
+    // On linux, the prof timer will deliver the signal to the thread which triggered the timer,
+    // because these timers are based on process and system time, and as such, are thread-aware.
+    // For the real timer, the signal gets delivered to the main thread, seemingly always.
+    // Consequently if we want to sample multiple threads, we need to forward this signal.
+    if (signal_type == SIGALRM) {
+        if (is_main_thread() && broadcast_signal_for_threads()) {
+            __sync_lock_release(&spinlock);
```

Comment:
Yes and no - it's used to protect whatever it was normally protecting *and* the threads array.

If we don't need to sample the main thread (when it's not registered for real-time sampling in the first place), we can bail out early here, because we just need to broadcast the signal - but we still need to release the spinlock before the immediate return.